src/main/xlat_func.c

```diff
@@ -658,20 +658,58 @@ static ssize_t rand_xlat(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
  *  Build strings of random chars, useful for generating tokens and passcodes
  *  Format similar to String::Random.
  */
-static ssize_t randstr_xlat(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
-			    UNUSED void const *mod_inst, UNUSED void const *xlat_inst,
-			    REQUEST *request, char const *fmt)
+static xlat_action_t xlat_randstr(TALLOC_CTX *ctx, fr_cursor_t *out,
+				  REQUEST *request, UNUSED void const *xlat_inst, UNUSED void *xlat_thread_inst,
+				  fr_value_box_t **in)
 {
 	char const 	*p;
-	char		*out_p = *out;
+	char		*endptr;
+	char		*buff, *buff_p;
 	unsigned int	result;
 	unsigned int	number;
-	size_t		freespace = outlen;
+	size_t		outlen = 0;
+	fr_value_box_t*	vb;
 
-	if (outlen <= 1) return 0;
+	/*
+	 * Nothing to do if input is empty
+	 */
+	if (!(*in)) {
+		return XLAT_ACTION_DONE;
+	}
 
-	p = fmt;
-	while (*p && (--freespace > 0)) {
+	/*
+	 * Concatenate all input
+	 */
+	if (fr_value_box_list_concat(ctx, *in, in, FR_TYPE_STRING, true) < 0) {
+		RPEDEBUG("Failed concatenating input");
+		return XLAT_ACTION_FAIL;
+	}
+
+	p = (*in)->vb_strvalue;
+
+	/*
+	 * Calculate size of output
+	 */
+	while (*p) {
```

Comment:
You can't assume the input string is binary safe anymore (though the original code did).

General pattern is declare `char const *end`

```
end = p + (*in)->vb_length;

while (p < end) {
<stuff>
}
```

Reply:
The fact we pass FR_TYPE_STRING when calling fr_value_box_list_concat is misleading then. I was expecting the resulting valuebox to contain a NULL terminated string.

Reply:
The string is actually '\0' terminated for safety, but you should use the length anyway so we can error out correctly if there's a '\0' somewhere in the format string... Which yes I know sounds pretty stupid, but there are all kinds of fun attacks you can do with functions which don't deal with embedded '\0s' correctly.  I remember about 10 years ago a subset of PHP regex functions were not binary safe, so if you didn't want the remainder of your input string to be validated against a regex, you just stuck a '\0' in there.